Doc/c-api/init.rst

```diff
@@ -919,8 +919,10 @@ from a C thread is::
 Note that the ``PyGILState_*`` functions assume there is only one global
 interpreter (created automatically by :c:func:`Py_Initialize`).  Python
 supports the creation of additional interpreters (using
-:c:func:`Py_NewInterpreter`), but mixing multiple interpreters and the
-``PyGILState_*`` API is unsupported.
+:c:func:`Py_NewInterpreter`), but switching between interpreters via the
+``PyGILState_*`` API is unsupported. On the other hand, *creating*
+sub-interpreters (whether they have a per-interpreter GIL or not, in
+the sense of :pep:`684`) still require to hold the GIL.
```

Comment:
I don't think the note about creating interpreters should be *here*.
```suggestion
``PyGILState_*`` API is unsupported.
```
I think it belongs to [`Py_NewInterpreterFromConfig`](https://docs.python.org/3.14/c-api/init.html#c.Py_NewInterpreterFromConfig) -- but, it turns out it's already there:

> Like all other Python/C API functions, the global interpreter lock must be held before calling this function and is still held when it returns.

---

Also, for future reference: PEPs become *historical documents* when they're finalized. It's OK to link to them for background information or for reasoning behind a change, but for the current state, you should link to docs. (That would have been [`PyInterpreterConfig.gil`](https://docs.python.org/3.14/c-api/init.html#c.PyInterpreterConfig.gil) in this case.)

---

Looking at what remains with my suggestion -- changing:
> mixing multiple interpreters and the ``PyGILState_*`` API is unsupported.

to:
> switching between interpreters via the ``PyGILState_*`` API is unsupported.

That would mean it's OK to use ``PyGILState_Ensure`` without the GIL held, but *while* multiple interpreters are active.
I don't think we want to call that supported, at least without clarifying (and testing) *which* intprpreter you end in after calling `PyGILState_Ensure`. Because if you end up in an unexpected one, things can get messy -- for starters, all your `PyObject*` references are invalid.

Reply:
> I don't think we want to call that supported, at least without clarifying (and testing) which intprpreter you end in after calling PyGILState_Ensure

I think that would be better fit for the `PyGILState_Ensure` documentation -- we only need to change things here because it kept getting misinterpreted by people using subinterpreters to get around the GIL.

Reply:
After some testing, I think the `Py_NewInterpreterFromConfig` docs are wrong?

It says that the GIL is still held when it returns, which I interpreted as "you still need to call `PyGILState_Release`." But it seems that using a per-interpreter GIL releases the old one, and then switches to the new one.

For example (error checking omitted):

```c
#include <Python.h>
#include <pthread.h>
#define THREADS 10

static PyInterpreterConfig config = {
    .allow_threads = 1,
    .check_multi_interp_extensions = 1,
    .gil = PyInterpreterConfig_OWN_GIL,
};

void *thread_start(void *arg)
{
    PyGILState_STATE gil = PyGILState_Ensure();
    PyThreadState *tstate = NULL;
    Py_NewInterpreterFromConfig(&tstate, &config);
    PyObject *code = Py_CompileString("print(1)", __FILE__, Py_eval_input);
    PyObject *globals = PyDict_New();
    PyEval_EvalCode(code, globals, globals);
    Py_EndInterpreter(tstate);
    return NULL;
}

int main(void)
{
    Py_Initialize();
    pthread_t threads[THREADS];

    for (int i = 0; i < THREADS; ++i)
        pthread_create(&threads[i], NULL, thread_start, NULL);

    Py_BEGIN_ALLOW_THREADS;
    for (int i = 0; i < THREADS; ++i)
        pthread_join(threads[i], NULL);
    Py_END_ALLOW_THREADS;

    Py_Finalize();
    return 0;
}
```

I would have expected the above to deadlock, but it seems to work just fine. In fact, releasing `gil` causes it to break.